File: GusfieldsAlgo.py
def GusfieldsAlgo(str, pattern):
    new_string = pattern + "#" + str
    Z_Array = [0] * len(new_string)

    left = 0
    right = 0

    for i in range(1, len(new_string)):
        if i > right:
            left, right = i, i
            while right < len(new_string) and new_string[right - left] == new_string[right]:
                right += 1
            Z_Array[i] = right - left
            right -= 1
        else:
            k = i - left
            if Z_Array[k] < right - i + 1:
                Z_Array[i] = Z_Array[k]
            else:
                left = i
                while right < len(new_string) and new_string[right-left] == new_string[right]:
                    right += 1
                Z_Array[i] = right - left
                right -= 1
    for i in range(len(new_string)):
        if Z_Array[i] == len(pattern):
            print("Pattern found!")
        

if __name__ == "__main__":
    print(GusfieldsAlgo("asdeasdsadHello, my name is Eden haa,asdwads", "Hello, my name is Eden haha"))

File: GusfieldsAlgo.mjs
function GusfieldsAlgo(str, pattern) {
    const new_string = pattern + "#" + str;
    const Z_Array = new Array(new_string.length).fill(0);

    let left = 0;
    let right = 0;

    for (let i = 1; i < new_string.length; i++) {
        if (i > right) {
            left = i;
            right = i;
            while (right < new_string.length && new_string[right - left] === new_string[right]) {
                right++;
            }
            Z_Array[i] = right - left;
            right--;
        } else {
            const k = i - left;
            if (Z_Array[k] < right - i + 1) {
                Z_Array[i] = Z_Array[k];
            } else {
                left = i;
                while (right < new_string.length && new_string[right - left] === new_string[right]) {
                    right++;
                }
                Z_Array[i] = right - left;
                right--;
            }
        }
    }
    for (let i = 0; i < new_string.length; i++) {
        if (Z_Array[i] === pattern.length) {
            console.log("Pattern found!");
        }
    }
}
export {GusfieldsAlgo}